data_iterator.py

```diff
@@ -51,22 +56,35 @@ def create_padded_batch(state, x):
             X[triple_length:, idx] = state['eos_sym']
 
         # Initialize Xmask column with ones in all positions that
-        # were just set in X
+        # were just set in X. 
+        # Note: if we need mask to depend on tokens inside X, then we need to 
+        # create a corresponding mask for X_reversed and send it further in the model
         Xmask[:triple_length, idx] = 1.
 
-        # Find start of last utterance
+        # Reverse all utterances
         eos_indices = numpy.where(X[:, idx] == state['eos_sym'])[0]
+        X_reversed[:triple_length, idx] = x[0][idx][:triple_length]
+        prev_eos_index = -1
+        for eos_index in eos_indices:
+            X_reversed[(prev_eos_index+2):eos_index, idx] = (X_reversed[(prev_eos_index+2):eos_index, idx])[::-1]
+            prev_eos_index = eos_index
+            if prev_eos_index > triple_length:
+                break
```

Comment:
I don't have any significant slowdowns. I suspect the main bottleneck in the code right now (e.g. >80% time) is transferring data back and forth between CPU and GPU, not actually loading in the data and preprocessing it. But we can profile it if you want?

Anyway, I didn't use a flag there because I think we probably need to use the backward utterances in different future models anyway. So you really want to make data_iterator as independent of the model as possible.